cpp/src/strings/convert/convert_durations.cu

```diff
@@ -0,0 +1,761 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/get_value.cuh>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/types.hpp>
+#include <strings/convert/utilities.cuh>
+#include <strings/utilities.cuh>
+
+#include <thrust/transform_reduce.h>
+#include <map>
+#include <rmm/device_uvector.hpp>
+#include <vector>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+namespace {
+
+// duration components timeparts structure
+struct alignas(4) duration_component {
+  int32_t day;        //-2,147,483,648 to 2,147,483,647
+  int32_t subsecond;  // 000000000 to 999999999
+  int8_t hour;        // 00 to 23
+  int8_t minute;      // 00 to 59
+  int8_t second;      // 00 to 59
+  bool is_negative;   // true/false
+};
+
+enum class format_char_type : int8_t {
+  literal,   // literal char type passed through
+  specifier  // duration format specifier
+};
+
+/**
+ * @brief Represents a format specifier or literal from a duration format string.
+ *
+ * Created by the format_compiler when parsing a format string.
+ */
+struct alignas(4) format_item {
+  format_char_type item_type;  // specifier or literal indicator
+  char value;                  // specifier or literal value
+  int8_t length;               // item length in bytes
+
+  static format_item new_specifier(char format_char, int8_t length)
+  {
+    return format_item{format_char_type::specifier, format_char, length};
+  }
+  static format_item new_delimiter(char literal)
+  {
+    return format_item{format_char_type::literal, literal, 1};
+  }
+};
+
+/**
+ * @brief The format_compiler parses a duration format string into a vector of
+ * format_items.
+ *
+ * The vector of format_items are used when parsing a string into duration
+ * components and when formatting a string from duration components.
+ */
+struct format_compiler {
```

Comment:
I think `format_compiler` can be a lot simpler with `std::regex`. Here's how: https://godbolt.org/z/6ffo8M

Note that the errors are not as detailed, but is full detail on the location of incorrect format specifiers really required?

The nice thing is that the regex for the format specifiers is very simple -- no fancy regex features like lookahead required.

Reply:
This is a nice idea but the regex example does not include handling the literals but just skips over them. The format compiler must account for all the characters in the format and not just the specifiers. Change the godbolt example to:
```
format_compiler comp("%D:%n:%OH");
```
and you will see the `:` characters are missing from the format items. Perhaps there is a way to fix this with the regex logic but I'm not seeing it.

Reply:
Handling the literals are here in the code
https://github.com/rapidsai/cudf/blob/0d84102978a1a84ebab507764e1a1bcf09d1c8cf/cpp/src/strings/convert/convert_durations.cu#L100-L103

Reply:
Ah, then I think there's a doc bug. The docs just say `The format pattern can include the following specifiers: "%Y,%y,%m,%d,%H,%I,%p,%M,%S,%f,%z"` -- Nothing about other characters or strings that can be in the format specifier.

Reply:
This regex will add the ability to capture individual characters in between the format specifiers: `\%([\%ntDHIMSpRTr]|OH|OI|OM|OS)|(.)`

(And this for the full match: `^(\%([\%ntDHIMSpRTr]|OH|OI|OM|OS)|(.))+$`)

https://regex101.com/r/LWqfGq/1

But if you need to capture arbitrary strings delimited by format specifiers it will be a little different. I don't think the docs fully specify it.

Reply:
I think this is closer? https://godbolt.org/z/n45c85

Output for `format_compiler comp("%D:%n:%OH");`

```
6
Item 0: - -1
Item 1: D -1
Item 2: : 1
Item 3: 
 1
Item 4: : 1
Item 5: H 2
```

Reply:
This is closer but does not recognize invalid specifiers correctly: `format_compiler comp("%D:%n:%OH:%q");`
I feel this getting more complicated than the original code.